bcachefs/bcachefs_iterator.c

```diff
@@ -434,14 +462,49 @@ int Bcachefs_iter_fini(const Bcachefs *this, Bcachefs_iterator *iter)
     iter->btree_node = NULL;
     free(iter->_bsets);
     iter->_bsets = NULL;
+    free(iter->_btree_ptrs);
+    iter->_btree_ptrs = NULL;
     *iter = (Bcachefs_iterator){
         .type = BTREE_ID_NR,
         .btree_node = iter->btree_node,
         .next_it = iter->next_it,
-        ._bsets = iter->_bsets
+        ._bsets = iter->_bsets,
+        ._btree_ptrs = iter->_btree_ptrs
     };
     return iter->next_it == NULL && iter->btree_node == NULL &&
-        iter->_bsets == NULL;
+        iter->_bsets == NULL && iter->_btree_ptrs == NULL;
+}
+
+void _Bcachefs_iter_append_bsets_btree_ptr(Bcachefs_iterator *iter, const struct bch_btree_ptr_v2 *btree_ptr)
+{
+    if (iter->_btree_ptrs_first == iter->_btree_ptrs)
+    {
+        const struct bch_btree_ptr_v2 **swp = (void*)iter->_btree_ptrs;;
+        const uint32_t num = iter->_btree_ptrs_end - (const struct bch_btree_ptr_v2**)iter->_btree_ptrs;
+        iter->_btree_ptrs = malloc(sizeof(struct bch_btree_ptr_v2*) * num * 2);
+        iter->_btree_ptrs_end = (void*)(iter->_btree_ptrs + num * 2);
+        iter->_btree_ptrs_first = iter->_btree_ptrs + num;
+        iter->_btree_ptr = iter->_btree_ptrs_end;
+        memcpy(iter->_btree_ptrs_first, swp, sizeof(struct bch_btree_ptr_v2*) * num);
```

Comment:
Minor nit: Remember that in C it's legal to do `sizeof(*iter->_btree_ptrs)` (note the `*` dereference) to get the sizeof of the type that it points to. Then you do not need to copy-paste the type all over the place. Obviously this only works if the type of the pointed type is, in fact, the type of the array elements (and not `void*`), but that's the case here.

Reply:
ah that's probably cleaner